State management

EddyVinck opened the pull request:
The Apollo docs about state management alone are giant. I'm trying to make it a bit more condensed with references to their docs while also showcasing how things work with Falcon. There still is a lot to cover, so I added a "short version" for people already familiar with Apollo Client.

Let me know if I forgot something about the Falcon side of things. I looked at the demo project and I couldn't find many deviations from the way Apollo Client works (which is a good thing, imo).

Preview: https://deploy-preview-70--falcon-docs.netlify.com/docs/falcon-client/local-state-management

Borales commented on docs/falcon-client/state-management.md:
```diff
@@ -2,4 +2,523 @@
+
+This will go in-depth about how to use Apollo for managing the client state of your Falcon application with Apollo.
+
+If youre unfamiliar with GraphQL queries it is recommended to check out [this guide](https://graphql.org/learn/queries/).
```
```suggestion
If you're unfamiliar with GraphQL queries it is recommended to check out [this guide](https://graphql.org/learn/queries/).
```

Borales commented on docs/falcon-client/state-management.md:
```diff
@@ -2,4 +2,523 @@
+
+The `Query` API is pretty simple if you're familiar with React and the [Render Prop](https://reactjs.org/docs/render-props.html) pattern.
+
+React will call the render prop function you provide with an object from Apollo Client containing loading, error, and data properties that you can use to render your UI.
```
```suggestion
React will call the render prop function you provide with an object from Apollo Client containing `loading`, `error`, and `data` properties that you can use to render your UI.
```

EddyVinck commented on docs/falcon-client/state-management.md:
```diff
@@ -2,4 +2,523 @@
+
+Falcon comes with a few pre-defined queries and mutations when you create a project with [`create-falcon-app`](getting-started/installation.md#create-falcon-app). These are primarily from our [`@deity/falcon-ecommerce-uikit`](https://github.com/deity-io/falcon/tree/master/packages/falcon-ecommerce-uikit) package and can be of great use when starting out with an eCommerce application with an integ ...
+
+A pattern you will see quite a bit in our included (but optional) components is to create a custom `Query` component with the query added as a default prop.
```
@Borales I'll try to rephrase it. It is kind of vague, I agree.

nimobeeren commented on docs/falcon-client/state-management.md:
```diff
@@ -2,4 +2,523 @@
+
+Your initialized local state and the resolvers go in `client/src/clientState.js`.
+
+### Falcon's additional components
```
Maybe you can also write something about Falcon's `Query` component (which extends Apollo's `Query`). The main features are displaying error and loading states. Unfortunately this is only useful when using the render prop method. But since this article uses that everywhere, it's useful to know about these features.

EddyVinck replied:
@nimobeeren good suggestion, I'll write a bit about that as well.

nimobeeren commented on docs/falcon-client/state-management.md:
```diff
@@ -2,4 +2,523 @@
+
+## The short version
+
+If you already know how to use Apollo Client, you pretty much already know how to use it with Falcon. If you need a refresher, continue to the [long version](falcon-client/state-management.md#the-long-version) or the [Apollo documentation](https://www.apollographql.com/docs/react/v2.5/essentials/get-started/).
```
```suggestion
If you already know how to use Apollo Client, you pretty much know how to use it with Falcon. If you need a refresher, continue to the [long version](falcon-client/state-management.md#the-long-version) or the [Apollo documentation](https://www.apollographql.com/docs/react/v2.5/essentials/get-started/).
```

nimobeeren commented on docs/falcon-client/state-management.md:
```diff
@@ -2,4 +2,523 @@
+
+## The long version
+
+This will go in-depth about how to use Apollo for managing the client state of your Falcon application with Apollo.
```
```suggestion
This will go in-depth about how to use Apollo for managing the client state of your Falcon application.
```

EddyVinck replied:
@nimobeeren I don't think that part should be removed, since most of it is about working with Apollo.

nimobeeren replied:
I agree about mentioning Apollo, but it is being mentioned two times in one sentence

EddyVinck replied:
oh, I totally missed that. Never mind :sweat_smile:

nimobeeren commented on docs/falcon-client/state-management.md:
````diff
@@ -2,4 +2,523 @@
+);
+```
+
+The `Query` component subscribes to the data through an observable. The result is first checked in the Apollo Client cache. If the result isn't in the cache it will make a request to the server. When the data is received from the server it will be stored in the cache.
````
```suggestion
The `Query` component subscribes to the data through an observable. The result is first checked in the Apollo Client cache. If the result isn't in the cache, it will make a request to the server. When the data is received from the server, it will be stored in the cache.
```

nimobeeren commented on docs/falcon-client/state-management.md:
```diff
@@ -2,4 +2,523 @@
+
+If you need near real-time data you can read more about that in the [Apollo documentation](https://www.apollographql.com/docs/react/v2.5/essentials/queries/#polling-and-refetching).
+
+If you need to notify the user when you are refetching data or if their network status has changes you can read more about that in the [Apollo documentation](https://www.apollographql.com/docs/react/v2.5/essentials/queries/#loading-and-error-state).
```
```suggestion
If you need to notify the user when you are refetching data or if their network status has changes, you can read more about that in the [Apollo documentation](https://www.apollographql.com/docs/react/v2.5/essentials/queries/#loading-and-error-state).
```

nimobeeren commented on docs/falcon-client/state-management.md:
```diff
@@ -2,4 +2,523 @@
+
+Apollo is great when using GraphQL to fetch data in your front-end, but you can also use Apollo Client to do local state management, just like you would use the [React Context API](https://reactjs.org/docs/context.html), MobX or Redux. You can use those other options as well, but using Apollo Client for local state management is ideal because your Apollo cache will be the single source of truth i ...
+
+You can read more about local state management with Apollo Client [here](https://www.apollographql.com/docs/react/v2.5/essentials/local-state/). Since Falcon Client does not support Apollo's hooks yet you should use the examples with Render Prop. You can use the dropdown above their code examples to change the examples from hooks to render props.
```
Maybe mention that `< 2.0.0` does not support hooks

nimobeeren commented on docs/falcon-client/state-management.md:
```diff
@@ -2,4 +2,523 @@
+
+#### Using and updating local client state
+
+When using local client state you need to write resolvers (queries and mutations) to access and edit the data. You can read more about that in Apollo's documentation about [local resolvers](https://www.apollographql.com/docs/react/v2.5/essentials/local-state/#local-resolvers).
```
```suggestion
When using local client state, you need to write resolvers to query and mutate the data. You can read more about that in Apollo's documentation about [local resolvers](https://www.apollographql.com/docs/react/v2.5/essentials/local-state/#local-resolvers).
```

Borales commented:
Since @EddyVinck is working on updating docs for Falcon 2.0 - I made another PR #71 to "pin" the current docs as "1.x" version, so we could proceed with updating docs for Falcon 2.0

EddyVinck commented:
@Borales But I'm not updating the docs for Falcon 2.0? 
I'm just updating the docs in general. As discussed in Slack I wasn't going to update the docs to  2.0 docs since I don't know the details as well as you guys do. 

I'm going to take a look at the suggestions for this PR today by the way.

EddyVinck commented:
@Borales I split the page into two pages and I changed the dog example to a product example. I'm not sure what to do with the to-do list example. I like how straightforward this to-do example is for the mutations and I'd like an easy to understand example if we want to replace it with something else. Any suggestions?

nimobeeren commented:
I think it's not necessary to include the "Additional Falcon components" section again on the local state management page. The Data Management page is more general so in my opinion it fits best there.

nimobeeren commented on docs/falcon-client/local-state-management.md:
````diff
@@ -0,0 +1,332 @@
+}
+```
+
+You can also use our wrapper `Query` component, which extends Apollo's `Query` component. The wrapper component will handle loading and error states for you. If you're curious how this works, you can check out the source code [here](). You can use our `Query` component in the same way you would use the Apollo Client version.
````
Source code link is missing

EddyVinck replied:
good catch!

EddyVinck commented:
> I think it's not necessary to include the "Additional Falcon components" section again on the local state management page. The Data Management page is more general so in my opinion it fits best there.

Agreed

Borales commented:
> I'm not sure what to do with the to-do list example. I like how straightforward this to-do example is for the mutations and I'd like an easy to understand example if we want to replace it with something else. Any suggestions?

@EddyVinck One of the Mutation examples we have is `addToCart` mutation. It not only changes the cache on the client-side, but it also calls `refetchQueries` to get the latest data on the Cart Summary for the UI (that we can't really do by ourselves) - I think we could use it for the example.

This is just a suggestion, we can leave the current to-do example, as it reflects a manual cache update and gives a better picture of how AC works.

EddyVinck commented on docs/falcon-client/data-management.md:
````diff
@@ -0,0 +1,282 @@
+}
+```
+
+You can also use our wrapper `Query` component, which extends Apollo's `Query` component. The wrapper component will handle loading and error states for you. If you're curious how this works, you can check out the source code [here](https://github.com/deity-io/falcon/blob/dev/packages/falcon-data/src/Query/Query.tsx). You can use our `Query` component in the same way you would use the Apollo Clie ...
````
@Borales Sure, sounds good to me!

EddyVinck commented:
> This is just a suggestion, we can leave the current to-do example, as it reflects a manual cache update and gives a better picture of how AC works.

@Borales I think the to-do example is fine, and to be honest I don't feel like rewriting it and digging in the code to change it to the `addToCart` example :see_no_evil: 
I merged all your textual suggestions. I think this PR can be merged now, unless you have additional feedback of course :slightly_smiling_face:

mateuszwozniak commented:
Great job with that! Thanks @EddyVinck!
